testsuite/tests/typing-layouts-unboxed-records/recursive.ml

```diff
@@ -106,92 +123,169 @@ type 'a bad = #{ bad : 'a bad ; u : 'a}
-         because it is an unboxed record.
-       But the layout of type "bad" must be a sublayout of float64
-         because of the annotation on the declaration of the type bad.
+Error: The definition of "bad" is recursive without boxing:
+         "bad" contains "bad"
+|}]
+
+type bad = #{ a : t ; b : t }
+[%%expect{|
+type bad = #{ a : t; b : t; }
+|}]
+
+type 'a bad = #{ a : 'a bad }
+[%%expect{|
+Line 1, characters 0-29:
+1 | type 'a bad = #{ a : 'a bad }
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The definition of "bad" is recursive without boxing:
+         "'a bad" contains "'a bad"
+|}]
+
+type bad = #( s * s )
+and ('a : any) record_id2 = #{ a : 'a }
+and s = #{ u : u }
+and u = #(int * bad record_id2)
+[%%expect{|
+Line 1, characters 0-21:
+1 | type bad = #( s * s )
+    ^^^^^^^^^^^^^^^^^^^^^
+Error: The definition of "bad" is recursive without boxing:
+         "bad" = "#(s * s)",
+         "#(s * s)" contains "u",
+         "u" = "#(int * bad record_id2)",
+         "#(int * bad record_id2)" contains "bad"
+|}]
+
+type bad = #( s * s )
+and ('a : any) record_id2 = #{ a : 'a }
+and s = #{ u : u }
+and u = #(int * bad alias_id record_id2)
+[%%expect{|
+Line 1, characters 0-21:
+1 | type bad = #( s * s )
+    ^^^^^^^^^^^^^^^^^^^^^
+Error: The definition of "bad" is recursive without boxing:
+         "bad" = "#(s * s)",
+         "#(s * s)" contains "u",
+         "u" = "#(int * bad alias_id record_id2)",
+         "#(int * bad alias_id record_id2)" contains "bad alias_id",
+         "bad alias_id" = "bad",
+         "bad" = "#(s * s)",
+         "#(s * s)" contains "s"
+|}]
+
+(* We also check recursive types via modules *)
+module rec Bad_rec1 : sig
+  type t = #( s * s )
+  and s = #{ u : Bad_rec2.u }
+end = struct
+  type t = #( s * s )
+  and s = #{ u : Bad_rec2.u }
+end
+and Bad_rec2 : sig
+  type u = Bad_rec1.t id
+  and 'a id = 'a
+end = struct
+  type u = Bad_rec1.t id
+  and 'a id = 'a
+end
+[%%expect{|
+Lines 1-7, characters 0-3:
+1 | module rec Bad_rec1 : sig
+2 |   type t = #( s * s )
+3 |   and s = #{ u : Bad_rec2.u }
+4 | end = struct
+5 |   type t = #( s * s )
+6 |   and s = #{ u : Bad_rec2.u }
+7 | end
+Error: The definition of "Bad_rec1.t" is recursive without boxing:
+         "Bad_rec1.t" = "#(Bad_rec1.s * Bad_rec1.s)",
+         "#(Bad_rec1.s * Bad_rec1.s)" contains "Bad_rec2.u",
+         "Bad_rec2.u" = "Bad_rec1.t Bad_rec2.id",
+         "Bad_rec1.t Bad_rec2.id" = "Bad_rec1.t",
+         "Bad_rec1.t" = "#(Bad_rec1.s * Bad_rec1.s)",
+         "#(Bad_rec1.s * Bad_rec1.s)" contains "Bad_rec1.s"
+|}]
+
+(* When we allow records with elements of unrepresentable layout, this should
+   still be disallowed. *)
+module M : sig
```

Comment:
This inspired a few more tests:

```ocaml
module M : sig
  type 'a t
end = struct
  type 'a t = 'a
end

type bad = #{ bad : bad M.t }
```

is accepted, even though it creates a bottomless type. This is harmless though because its size is finite.

But I also tried

```ocaml
module M : sig
  type ('a : value & value) t : value & value
end = struct
  type ('a : any) t = 'a
end

type bad : value & value = #{ x : int; bad : bad M.t }
```

which leads to

```
Error: The layout of type bad is value & (value & value)
         because it is an unboxed record.
       But the layout of type bad must be a sublayout of value & value
         because of the annotation on the declaration of the type bad.
```

This looks like the old message this patch is an attempt to improve. (Without the `value & value` annotation, the error message mentions `any & any`.)

I don't see an actual problem here, but I do think this example shows that we're still relying on the previous behavior for detecting these loops, at least for devious users.